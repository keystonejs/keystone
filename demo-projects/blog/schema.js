require('dotenv').config();
const {
  File,
  Text,
  Slug,
  Relationship,
  Select,
  Password,
  Checkbox,
  CalendarDay,
  DateTime,
  OEmbed,
} = require('@keystonejs/fields');
const { Wysiwyg } = require('@keystonejs/fields-wysiwyg-tinymce');
const { LocalFileAdapter } = require('@keystonejs/file-adapters');
const getYear = require('date-fns/get_year');

const { staticRoute, staticPath, distDir } = require('./config');
const dev = process.env.NODE_ENV !== 'production';

let iframelyAdapter;

if (process.env.IFRAMELY_API_KEY) {
  const { IframelyOEmbedAdapter } = require('@keystonejs/oembed-adapters');
  iframelyAdapter = new IframelyOEmbedAdapter({
    apiKey: process.env.IFRAMELY_API_KEY,
  });
}

const fileAdapter = new LocalFileAdapter({
  src: `${dev ? '' : `${distDir}/`}${staticPath}/uploads`,
  path: `${staticRoute}/uploads`,
});

const avatarFileAdapter = new LocalFileAdapter({
  src: `${staticPath}/avatars`,
  path: `${staticRoute}/avatars`,
});

exports.User = {
  fields: {
    name: { type: Text },
    email: { type: Text, isUnique: true },
    dob: {
      type: CalendarDay,
      format: 'Do MMMM YYYY',
      yearRangeFrom: 1901,
      yearRangeTo: getYear(new Date()),
    },
    ...(process.env.IFRAMELY_API_KEY
      ? {
          portfolio: { type: OEmbed, adapter: iframelyAdapter },
        }
      : {}),
    password: { type: Password },
    isAdmin: { type: Checkbox },
    avatar: { type: File, adapter: avatarFileAdapter },
  },
  labelResolver: item => `${item.name} <${item.email}>`,
};

exports.Post = {
  fields: {
    title: { type: Text },
    slug: { type: Slug, from: 'title' },
    author: {
      type: Relationship,
      ref: 'User',
    },
    categories: {
      type: Relationship,
      ref: 'PostCategory',
      many: true,
    },
    status: {
      type: Select,
      defaultValue: 'draft',
      options: [
        { label: 'Draft', value: 'draft' },
        { label: 'Published', value: 'published' },
      ],
    },
    body: { type: Wysiwyg },
    posted: { type: DateTime, format: 'DD/MM/YYYY' },
    image: { type: File, adapter: fileAdapter },
  },
  adminConfig: {
    defaultPageSize: 20,
    defaultColumns: 'title, status',
    defaultSort: 'title',
  },
  labelResolver: item => item.title,
};

exports.PostCategory = {
  fields: {
    name: { type: Text },
    slug: { type: Slug, from: 'name' },
  },
};

exports.Comment = {
  fields: {
    body: { type: Text, isMultiline: true },
    originalPost: {
      type: Relationship,
      ref: 'Post',
    },
    author: {
      type: Relationship,
      ref: 'User',
    },
    posted: { type: DateTime },
  },
  labelResolver: item => item.body,
};
