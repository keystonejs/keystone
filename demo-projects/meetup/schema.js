require('dotenv').config();
const uuid = require('uuid/v4');
const { sendEmail } = require('./emails');

const {
  CloudinaryImage,
  Checkbox,
  DateTime,
  Integer,
  Password,
  Relationship,
  Select,
  Text,
} = require('@keystonejs/fields');
const { CloudinaryAdapter } = require('@keystonejs/file-adapters');
const { Wysiwyg } = require('@keystonejs/fields-wysiwyg-tinymce');

const cloudinaryAdapter = new CloudinaryAdapter({
  cloudName: process.env.CLOUDINARY_CLOUD_NAME,
  apiKey: process.env.CLOUDINARY_KEY,
  apiSecret: process.env.CLOUDINARY_SECRET,
});

const access = {
  userIsAdmin: ({ authentication: { item: user } }) => Boolean(user && user.isAdmin),
  userIsCurrentAuth: ({ authentication: { item: user } }) => Boolean(user), // item will be undefined for anonymous user
};

// Read: public / Write: admin
const DEFAULT_LIST_ACCESS = {
  create: access.userIsAdmin,
  read: true,
  update: access.userIsAdmin,
  delete: access.userIsAdmin,
};

exports.User = {
  access: {
    update: access.userIsCurrentAuth,
    delete: access.userIsAdmin,
  },
  fields: {
    name: { type: Text },
    email: { type: Text, isUnique: true, access: { read: access.userIsCurrentAuth } },
    password: { type: Password, isRequired: true },
    isAdmin: { type: Checkbox, access: { update: access.userIsAdmin } },
    twitterHandle: { type: Text },
    image: { type: CloudinaryImage, adapter: cloudinaryAdapter },
    talks: {
      type: Relationship,
      ref: 'Talk.speakers',
      many: true,
      access: { update: access.userIsAdmin },
    },
  },
  hooks: {
    afterChange: async ({ updatedItem, existingItem }) => {
      if (existingItem && updatedItem.password !== existingItem.password) {
        const url = process.env.SERVER_URL || 'http://localhost:3000';

        const props = {
          recipientEmail: updatedItem.email,
          signinUrl: `${url}/signin`,
        };

        const options = {
          subject: 'Your password has been updated',
          to: updatedItem,
          from: process.env.MAILGUN_FROM,
          domain: process.env.MAILGUN_DOMAIN,
          apiKey: process.env.MAILGUN_API_KEY,
        };

        await sendEmail('password-updated.jsx', props, options);
      }
    },
  },
};

exports.Organiser = {
  access: DEFAULT_LIST_ACCESS,
  fields: {
    user: { type: Relationship, ref: 'User' },
    order: { type: Integer },
    role: { type: Text },
  },
};

// TODO: We can't access the existing item at the list update level yet,
// read access needs to check if event is "active" or if the user is admin
// read: ({ existingItem, authentication }) => access.userIsAdmin({ authentication }) || !!(existingItem && existingItem.status === 'active'),

exports.Event = {
  access: DEFAULT_LIST_ACCESS,
  fields: {
    name: { type: Text },
    status: { type: Select, options: 'draft, active', defaultValue: 'draft' },
    themeColor: { type: Text },
    startTime: { type: DateTime },
    durationMins: { type: Integer },
    description: { type: Wysiwyg },
    talks: { type: Relationship, ref: 'Talk.event', many: true },
    locationAddress: { type: Text },
    locationDescription: { type: Text },
    maxRsvps: { type: Integer, defaultValue: 120 },
    isRsvpAvailable: { type: Checkbox, defaultValue: true },
  },
};

exports.Talk = {
  access: DEFAULT_LIST_ACCESS,
  fields: {
    name: { type: Text },
    event: { type: Relationship, ref: 'Event.talks' },
    speakers: { type: Relationship, ref: 'User.talks', many: true },
    isLightningTalk: { type: Checkbox },
    description: { type: Wysiwyg },
  },
};

exports.Rsvp = {
  access: {
    create: true,
    read: true,
    update: ({ authentication: { item } }) => {
      if (!item) {
        return false;
      }
      return { user: { id: item.id } };
    },
    delete: access.userIsAdmin,
  },
  fields: {
    event: { type: Relationship, ref: 'Event' },
    user: { type: Relationship, ref: 'User' },
    status: { type: Select, options: 'yes, no' },
  },
  hooks: {
    validateInput: async ({ resolvedData, existingItem, actions }) => {
      const { status } = resolvedData;
      const { event: eventId } = existingItem ? existingItem : resolvedData;

      if (status === 'no') {
        return;
      }

      const { data } = await actions.query(`query {
        event: Event(where: { id: "${eventId}" }) {
          id
          startTime
          maxRsvps
          isRsvpAvailable
        }
        allRsvps(where: { event: { id: "${eventId}" }}) {
          id
        }
      }`);

      const { event, allRsvps } = data;

      if (
        !event ||
        !event.isRsvpAvailable ||
        !event.startTime ||
        new Date() > new Date(event.startTime) ||
        allRsvps.length >= event.maxRsvps
      ) {
        throw 'Error rsvping to event';
      }
    },
  },
};

exports.Sponsor = {
  access: DEFAULT_LIST_ACCESS,
  fields: {
    name: { type: Text },
    website: { type: Text },
    logo: { type: CloudinaryImage, adapter: cloudinaryAdapter },
  },
};

exports.ForgottenPasswordToken = {
  access: {
    create: true,
    read: true,
    update: access.userIsAdmin,
    delete: access.userIsAdmin,
  },
  fields: {
    user: {
      type: Relationship,
      ref: 'User',
      access: {
        read: access.userIsAdmin,
      },
    },
    token: {
      type: Text,
      isRequired: true,
      isUnique: true,
      access: {
        read: access.userIsAdmin,
      },
    },
    requestedAt: { type: DateTime, isRequired: true },
    accessedAt: { type: DateTime },
    expiresAt: { type: DateTime, isRequired: true },
  },
  hooks: {
    afterChange: async ({ updatedItem, existingItem, actions: { query } }) => {
      if (existingItem) return null;

      const now = new Date().toISOString();

      const { errors, data } = await query(
        `
        query GetUserAndToken($user: ID!, $now: DateTime!) {
          User( where: { id: $user }) {
            id
            email
          }
          allForgottenPasswordTokens( where: { user: { id: $user }, expiresAt_gte: $now }) {
            token
            expiresAt
          }
        }
      `,
        { skipAccessControl: true, variables: { user: updatedItem.user.toString(), now } }
      );

      if (errors) {
        console.error(errors, `Unable to construct password updated email.`);
        return;
      }

      const { allForgottenPasswordTokens, User } = data;
      const forgotPasswordKey = allForgottenPasswordTokens[0].token;
      const url = process.env.SERVER_URL || 'http://localhost:3000';

      const props = {
        forgotPasswordUrl: `${url}/change-password?key=${forgotPasswordKey}`,
        recipientEmail: User.email,
      };

      const options = {
        subject: 'Request for password reset',
        to: User.email,
        from: process.env.MAILGUN_FROM,
        domain: process.env.MAILGUN_DOMAIN,
        apiKey: process.env.MAILGUN_API_KEY,
      };

      await sendEmail('forgot-password.jsx', props, options);
    },
  },
  mutations: [
    {
      schema: 'startPasswordRecovery(email: String!): ForgottenPasswordToken',
      resolver: async (obj, { email }, context, info, { query }) => {
        const token = uuid();

        const tokenExpiration =
          parseInt(process.env.RESET_PASSWORD_TOKEN_EXPIRY) || 1000 * 60 * 60 * 24;

        const now = Date.now();
        const requestedAt = new Date(now).toISOString();
        const expiresAt = new Date(now + tokenExpiration).toISOString();

        const { errors: userErrors, data: userData } = await query(
          `
            query findUserByEmail($email: String!) {
              allUsers(where: { email: $email }) {
                id
                email
              }
            }
          `,
          { variables: { email: email }, skipAccessControl: true }
        );

        if (userErrors) {
          console.error(
            userErrors,
            `Unable to find user when trying to create forgotten password token.`
          );
          return;
        }

        const userId = userData.allUsers[0].id;

        const result = {
          userId,
          token,
          requestedAt,
          expiresAt,
        };

        const { errors } = await query(
          `
            mutation createForgottenPasswordToken(
              $userId: ID!,
              $token: String,
              $requestedAt: DateTime,
              $expiresAt: DateTime,
            ) {
              createForgottenPasswordToken(data: {
                user: { connect: { id: $userId }},
                token: $token,
                requestedAt: $requestedAt,
                expiresAt: $expiresAt,
              }) {
                id
                token
                user {
                  id
                }
                requestedAt
                expiresAt
              }
            }
          `,
          { variables: result, skipAccessControl: true }
        );

        if (errors) {
          console.error(errors, `Unable to create forgotten password token.`);
          return;
        }

        return true;
      },
    },
    {
      schema: 'changePasswordWithToken(token: String!, password: String!): User',
      resolver: async (obj, { token, password }, context, info, { query }) => {
        const now = Date.now();

        const { errors, data } = await query(
          `
            query findUserFromToken($token: String!, $now: DateTime!) {
              passwordTokens: allForgottenPasswordTokens(where: { token: $token, expiresAt_gte: $now }) {
                id
                token
                user {
                  id
                }
              }
            }
          `,
          { variables: { token, now }, skipAccessControl: true }
        );

        if (errors || !data.passwordTokens || !data.passwordTokens.length) {
          console.error(errors, `Unable to find token`);
          throw errors.message;
        }

        const user = data.passwordTokens[0].user.id;
        const tokenId = data.passwordTokens[0].id;

        const { errors: passwordError } = await query(
          `mutation UpdateUserPassword($user: ID!, $password: String!) {
            updateUser(id: $user, data: { password: $password }) {
              id
            }
          }
        `,
          { variables: { user, password }, skipAccessControl: true }
        );

        if (passwordError) {
          console.error(passwordError, `Unable to change password`);
          throw passwordError.message;
        }

        await query(
          `mutation DeletePasswordToken($tokenId: ID!) {
            deleteForgottenPasswordToken(id: $tokenId) {
              id
            }
          }
        `,
          { variables: { tokenId }, skipAccessControl: true }
        );

        return true;
      },
    },
  ],
};
