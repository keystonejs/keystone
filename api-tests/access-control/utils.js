const { setupServer } = require('@keystonejs/test-utils');
const { Text } = require('@keystonejs/fields');
const { objMerge } = require('@keystonejs/utils');

const FAKE_ID = { mongoose: '5b3eabd9e9f2e3e4866742ea', knex: 137 };
const FAKE_ID_2 = { mongoose: '5b3eabd9e9f2e3e4866742eb', knex: 138 };

const yesNo = truthy => (truthy ? 'Yes' : 'No');

const getPrefix = access => {
  // prettier-ignore
  let prefix = `${yesNo(access.create)}Create${yesNo(access.read)}Read${yesNo(access.update)}Update${yesNo(access.auth)}Auth`;
  if (Object.prototype.hasOwnProperty.call(access, 'delete')) {
    prefix = `${prefix}${yesNo(access.delete)}Delete`;
  }
  return prefix;
};

const getStaticListName = access => `${getPrefix(access)}StaticList`;
const getImperativeListName = access => `${getPrefix(access)}ImperativeList`;
const getDeclarativeListName = access => `${getPrefix(access)}DeclarativeList`;

/* Generated with:
  const result = [];
  const options = ['create', 'read', 'update', 'delete', 'auth'];
  // All possible combinations are contained in the set 0..2^n-1
  for(let flags = 0; flags < Math.pow(2, options.length); flags++) {
    // Generate an object of true/false values for the particular combination
    result.push(options.reduce((memo, option, index) => ({
      ...memo,
      // Use a bit mask to see if that bit is set
      [option]: !!(flags & (1 << index)),
    }), {}));
  }
  */
const listAccessVariations = [
  { create: false, read: false, update: false, delete: false, auth: true },
  { create: true, read: false, update: false, delete: false, auth: true },
  { create: false, read: true, update: false, delete: false, auth: true },
  { create: true, read: true, update: false, delete: false, auth: true },
  { create: false, read: false, update: true, delete: false, auth: true },
  { create: true, read: false, update: true, delete: false, auth: true },
  { create: false, read: true, update: true, delete: false, auth: true },
  { create: true, read: true, update: true, delete: false, auth: true },
  { create: false, read: false, update: false, delete: true, auth: true },
  { create: true, read: false, update: false, delete: true, auth: true },
  { create: false, read: true, update: false, delete: true, auth: true },
  { create: true, read: true, update: false, delete: true, auth: true },
  { create: false, read: false, update: true, delete: true, auth: true },
  { create: true, read: false, update: true, delete: true, auth: true },
  { create: false, read: true, update: true, delete: true, auth: true },
  { create: true, read: true, update: true, delete: true, auth: true },
  { create: false, read: false, update: false, delete: false, auth: false },
  { create: true, read: false, update: false, delete: false, auth: false },
  { create: false, read: true, update: false, delete: false, auth: false },
  { create: true, read: true, update: false, delete: false, auth: false },
  { create: false, read: false, update: true, delete: false, auth: false },
  { create: true, read: false, update: true, delete: false, auth: false },
  { create: false, read: true, update: true, delete: false, auth: false },
  { create: true, read: true, update: true, delete: false, auth: false },
  { create: false, read: false, update: false, delete: true, auth: false },
  { create: true, read: false, update: false, delete: true, auth: false },
  { create: false, read: true, update: false, delete: true, auth: false },
  { create: true, read: true, update: false, delete: true, auth: false },
  { create: false, read: false, update: true, delete: true, auth: false },
  { create: true, read: false, update: true, delete: true, auth: false },
  { create: false, read: true, update: true, delete: true, auth: false },
  { create: true, read: true, update: true, delete: true, auth: false },
];

const fieldMatrix = [
  { create: false, read: false, update: false },
  { create: true, read: false, update: false },
  { create: false, read: true, update: false },
  { create: true, read: true, update: false },
  { create: false, read: false, update: true },
  { create: true, read: false, update: true },
  { create: false, read: true, update: true },
  { create: true, read: true, update: true },
];
const getFieldName = access => getPrefix(access);

const nameFn = { imperative: getImperativeListName, declarative: getDeclarativeListName };

const createFieldStatic = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: fieldAccess,
  },
});
const createFieldImperative = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: {
      create: () => fieldAccess.create,
      read: () => fieldAccess.read,
      update: () => fieldAccess.update,
    },
  },
});
function setupKeystone(adapterName) {
  return setupServer({
    adapterName,
    createLists: keystone => {
      keystone.createList('User', { fields: { name: { type: Text } } });

      listAccessVariations.forEach(access => {
        keystone.createList(getStaticListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldStatic(variation))),
          },
          access,
        });
        keystone.createList(getImperativeListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldImperative(variation))),
          },
          access: {
            create: () => access.create,
            read: () => access.read,
            update: () => access.update,
            delete: () => access.delete,
            auth: () => access.auth,
          },
        });
        keystone.createList(getDeclarativeListName(access), {
          fields: { name: { type: Text } },
          access: {
            create: access.create,
            // arbitrarily restrict the data to a single item (see data.js)
            read: () => access.read && { name_starts_with: 'Hello' },
            update: () => access.update && { name_starts_with: 'Hello' },
            delete: () => access.delete && { name_starts_with: 'Hello' },
            auth: access.auth,
          },
        });
      });
    },
  });
}
module.exports = {
  FAKE_ID,
  FAKE_ID_2,
  getStaticListName,
  getImperativeListName,
  getDeclarativeListName,
  listAccessVariations,
  fieldMatrix,
  nameFn,
  setupKeystone,
  getFieldName,
};
