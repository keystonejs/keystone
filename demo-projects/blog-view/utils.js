const terminalLink = require('terminal-link');
const ciInfo = require('ci-info');
const chalk = require('chalk');

const ttyLink = (text, path, port) => {
  if (ciInfo.isCI) {
    return;
  }
  const url = `http://localhost:${port}${path}`;
  const link = terminalLink(url, url, { fallback: () => url });
  console.log(`ðŸ”— ${chalk.green(text)}\t${link}`);
};

function extractAppMeta(apps) {
  let adminPath;
  let graphiqlPath;
  let apiPath;

  apps.forEach(app => {
    switch (app.constructor.name) {
      case 'AdminUIApp': {
        adminPath = app.adminPath;
        break;
      }
      case 'GraphQLApp': {
        apiPath = app._apiPath;
        graphiqlPath = app._graphiqlPath;
        break;
      }
    }
  });

  return {
    adminPath,
    graphiqlPath,
    apiPath,
  };
}

function logAdminRoutes(apps, port) {
  const { adminPath, graphiqlPath, apiPath } = extractAppMeta(apps);
  /* eslint-disable no-unused-expressions */
  adminPath && ttyLink('Keystone Admin UI:', adminPath, port);
  graphiqlPath && ttyLink('GraphQL Playground:', graphiqlPath, port);
  apiPath && ttyLink('GraphQL API:\t', apiPath, port);
  /* eslint-enable no-unused-expressions */
}

module.exports = {
  logAdminRoutes,
};
