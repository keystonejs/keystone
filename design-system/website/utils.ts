export const toLabel = (str: string) => str.substr(0, 1).toUpperCase() + str.substr(1);

const VOWELS = ['a', 'e', 'i', 'o', 'u'];
export const aAn = (before: string) => `a${VOWELS.includes(before.charAt(0)) ? 'n' : ''}`;
