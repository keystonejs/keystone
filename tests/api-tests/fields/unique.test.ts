import globby from 'globby';
import { createSchema, list } from '@keystone-next/keystone/schema';
import { text } from '@keystone-next/fields';
import { setupTestEnv, setupTestRunner } from '@keystone-next/testing';
import { apiTestConfig, expectInternalServerError } from '../utils';

const testModules = globby.sync(`packages/**/src/**/test-fixtures.{js,ts}`, {
  absolute: true,
});
testModules
  .map(require)
  .filter(mod => !mod.skipUniqueTest)
  .filter(
    ({ supportsUnique, unSupportedAdapterList = [] }) =>
      supportsUnique && !unSupportedAdapterList.includes(process.env.TEST_ADAPTER)
  )
  .forEach(mod => {
    (mod.testMatrix || ['default']).forEach((matrixValue: string) => {
      describe(`${mod.name} - ${matrixValue} - isUnique`, () => {
        beforeEach(() => {
          if (mod.beforeEach) {
            mod.beforeEach();
          }
        });
        afterEach(async () => {
          if (mod.afterEach) {
            await mod.afterEach();
          }
        });
        beforeAll(() => {
          if (mod.beforeAll) {
            mod.beforeAll();
          }
        });
        afterAll(async () => {
          if (mod.afterAll) {
            await mod.afterAll();
          }
        });
        const runner = setupTestRunner({
          config: apiTestConfig({
            lists: createSchema({
              Test: list({
                fields: {
                  name: text(),
                  testField: mod.typeFunction({
                    isUnique: true,
                    ...(mod.fieldConfig ? mod.fieldConfig(matrixValue) : {}),
                  }),
                },
              }),
            }),
            images: { upload: 'local', local: { storagePath: 'tmp_test_images' } },
            files: { upload: 'local', local: { storagePath: 'tmp_test_files' } },
          }),
        });
        test(
          'uniqueness is enforced over multiple mutations',
          runner(async ({ context, graphQLRequest }) => {
            await context.lists.Test.createOne({
              data: { testField: mod.exampleValue(matrixValue) },
            });

            const { body } = await graphQLRequest({
              query: `
                  mutation($data: TestCreateInput) {
                    createTest(data: $data) { id }
                  }
                `,
              variables: { data: { testField: mod.exampleValue(matrixValue) } },
            });
            expect(body.data).toEqual({ createTest: null });
            expectInternalServerError(body.errors, [
              {
                path: ['createTest'],
                message: expect.stringMatching(
                  /\nInvalid `list\.listKey\).create\(\)` invocation in\n(.*\n)\n(.*\n){4}  Unique constraint failed on the fields: \(`testField`\)/
                ),
              },
            ]);
          })
        );

        test(
          'uniqueness is enforced over single mutation',
          runner(async ({ graphQLRequest }) => {
            const { body } = await graphQLRequest({
              query: `
                  mutation($fooData: TestCreateInput, $barData: TestCreateInput) {
                    foo: createTest(data: $fooData) { id }
                    bar: createTest(data: $barData) { id }
                  }
                `,
              variables: {
                fooData: { testField: mod.exampleValue(matrixValue) },
                barData: { testField: mod.exampleValue(matrixValue) },
              },
            });

            expect(body.data.foo).not.toBe(null);
            expect(body.data.bar).toBe(null);
            expectInternalServerError(body.errors, [
              {
                path: ['bar'],
                message: expect.stringMatching(
                  /\nInvalid `list\.listKey\).create\(\)` invocation in\n(.*\n)\n(.*\n){4}  Unique constraint failed on the fields: \(`testField`\)/
                ),
              },
            ]);
          })
        );

        test(
          'Configuring uniqueness on one field does not affect others',
          runner(async ({ context }) => {
            const items = await context.lists.Test.createMany({
              data: [
                { data: { testField: mod.exampleValue(matrixValue), name: 'jess' } },
                { data: { testField: mod.exampleValue2(matrixValue), name: 'jess' } },
              ],
            });
            expect(items).toHaveLength(2);
          })
        );
      });
    });
  });

testModules
  .map(require)
  .filter(mod => !mod.skipUniqueTest)
  .filter(
    ({ supportsUnique, unSupportedAdapterList = [] }) =>
      !supportsUnique &&
      supportsUnique !== null &&
      !unSupportedAdapterList.includes(process.env.TEST_ADAPTER)
  )
  .forEach(mod => {
    (mod.testMatrix || ['default']).forEach((matrixValue: string) => {
      describe(`${mod.name} - ${matrixValue} - isUnique`, () => {
        test('Ensure non-supporting fields throw an error', async () => {
          // Try to create a thing and have it fail
          let erroredOut = false;
          try {
            await setupTestEnv({
              config: apiTestConfig({
                lists: createSchema({
                  Test: list({
                    fields: {
                      name: text(),
                      testField: mod.typeFunction({
                        isUnique: true,
                        ...(mod.fieldConfig ? mod.fieldConfig(matrixValue) : {}),
                      }),
                    },
                  }),
                }),
                images: { upload: 'local', local: { storagePath: 'tmp_test_images' } },
                files: { upload: 'local', local: { storagePath: 'tmp_test_files' } },
              }),
            });
          } catch (error) {
            expect(error.message).toMatch('isUnique is not a supported option for field type');
            erroredOut = true;
          }
          expect(erroredOut).toEqual(true);
        });
      });
    });
  });
