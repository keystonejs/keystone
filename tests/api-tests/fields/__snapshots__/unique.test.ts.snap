// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigInt - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`BigInt - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`BigInt with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`BigInt with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Float - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Float - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Float with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Float with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Integer - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Integer - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Integer with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Integer with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - enum - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - enum - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - integer - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - integer - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - string - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select - string - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - enum - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - enum - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - integer - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - integer - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - string - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Select with isNullable: false - string - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Text - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Text - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Text with isNullable: true - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Text with isNullable: true - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Timestamp - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Timestamp - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Timestamp with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`Timestamp with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`calendarDay - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`calendarDay - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`calendarDay with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over multiple mutations 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;

exports[`calendarDay with isNullable: false - default - isIndexed: 'unique' uniqueness is enforced over single mutation 1`] = `
[
  [GraphQLError: Prisma error: Unique constraint failed on the fields: (\`testField\`)],
]
`;
